### ISSUE TITLE
Rule proposal: `indent` for template literals

### ISSUE THREAD

**Issue by fregante** [sindresorhus/eslint-plugin-unicorn#580]
This looks fine:

<img width="298" alt="ok" src="https://user-images.githubusercontent.com/1402241/76078551-f278a900-5fb3-11ea-8b70-a046b1ae49ef.png">

But is it?

<img width="301" alt="ko" src="https://user-images.githubusercontent.com/1402241/76078560-f4426c80-5fb3-11ea-9fc6-2a55d3106640.png">

## Description

Template literals are awkward because their indentation may be significant, but I think:

1. We have tools to make indentation at least partially insignificant, like https://github.com/sindresorhus/strip-indent
2. If leading/trailing whitespace is significant, it shouldn't be invisible. Use `\n` and `\t` if you _really_ want spaces in the output.

The second point could be an **option** that still forces indentation that's removable by `strip-indent` while still allowing _invisible but significant_ whitespace.

### Good 👍 

```js
function a () {
	return `
		Hello
	`;
}

function a () {
	return `
	  Hello
	`; // Mixed spacing *may* be ok
}
```

### Not good 👎 

```js
function a () {
	return `
	Hello
	`; // Wrong expected indentation of 'Hello'
}

function a () {
	return `
		Hello`; // Wrong position of closing "`"
}

function a () {
	return `
    Hello
    `; // 4 spaces instead of tab
}
```

## Related

Trailing spaces should also be dropped, but perhaps as a separate rule.

**Comment by fregante** [comment]
I'll add: The indentation in my example cannot be fully fixed because the parser has no concept of opening/closing tags **inside** the template literal, so:

```js
const a = `
••••hello {
••••••}
`;
```

can only be fixed to

```js
const a = `
→hello {
→••}
`;
```

or

```js
const a = `
→hello {
→→}
`;
```

and **not** as the expected:


```js
const a = `
→hello {
→}
`;
```

However:

1. The first erro is still useful
2. The second part can still be an **optional** error, e.g. "unexpected significant whitespace in template literal"

Also, **Note**: this may not be autofixable.

**Comment by sindresorhus** [comment]
Relevant: https://github.com/tc39/proposal-string-dedent

**Comment by bennypowers** [comment]
This is very relevant to lit-html, FAST, hybrids, html, and other templating systems

**Comment by karlhorky** [comment]
This would also be interesting in combination with tagged template literals containing SQL (eg. when using Postgres.js or Slonik)

**Comment by ArnaudBarre** [comment]
I don't how it works with non graphql templates, but for small project where I uses graphql with the whole gql library, I use this to trick prettier and get a nice formatting: `export const gql = (strings: TemplateStringsArray) => strings[0];`

**Comment by fisker** [comment]
Prettier support some well know tag functions like `` gql`code` ``, and comments like ``/* HTML */ `<html/>` ``, but we are not adding more languages, for plain text, there is an open issue https://github.com/prettier/prettier/issues/8207.

**Comment by mmkal** [comment]
@fisker @sindresorhus would a PR implementing this be accepted? If not, what would it take to get this from `:evaluating:` to `:help wanted:`?

**Comment by fisker** [comment]
I think we can accept this, my idea for the options, we accept 3 ways to target the template,

1, `tags`

By default, it will check well-known tags including `dedent`, `outdent` etc

2, `comments`

This targets ``/* HTML */ `...`;``, exactly like Prettier, the comment must be a block comment, and should appear before the template, even  parens  not  allowed

 ```js
const html = /* HTML */ (`...`); // No, won't match
```

```js
const html = (/* HTML */ `...`); // Yes
```

3, `selectors`

To target `stripIndent`'s first argument, use `CallExpression[callee.name="stripIndent"] > .argument:first-child`


To match any temple string , use `*`(we'll check node type anyway, so it's fine to select all) or  `TemplateLiteral`

**Comment by npdev453** [comment]
> This would also be interesting in combination with tagged template literals containing SQL (eg. when using Postgres.js or Slonik)

Yeah! `unicorn/template-indent` and Gajus's `sql/format` combo working pretty nice for now

https://github.com/gajus/eslint-plugin-sql/pull/33#issuecomment-1660748030

**Comment by karlhorky** [comment]
@npdev453 amazing! Would it be possible to make a video or 2 like I [suggested over here](https://github.com/gajus/eslint-plugin-sql/pull/33#issuecomment-1665225536)?

I'm guessing if more users knew about this, SQL formatting in tagged template literals in JS/TS would be a very desired feature (there has historically been no great, working option for this).

**Comment by npdev453** [comment]
> @npdev453 amazing! Would it be possible to make a video or 2 like I [suggested over here](https://github.com/gajus/eslint-plugin-sql/pull/33#issuecomment-1665225536)?
> 
> I'm guessing if more users knew about this, SQL formatting in tagged template literals in JS/TS would be a very desired feature (there has historically been no great, working option for this).

Done! https://youtu.be/9bZgvHfSnrQ

[![IYouTube](https://img.youtube.com/vi/9bZgvHfSnrQ/0.jpg)](https://youtu.be/9bZgvHfSnrQ)

**Comment by karlhorky** [comment]
For anyone finding this and looking for another alternative for formatting SQL in embedded `sql` tagged template literals:

Prettier can be used with the combination of the `prettier-plugin-embed` and `prettier-plugin-sql` plugins:

- https://github.com/Sec-ant/prettier-plugin-embed/issues/20
